torch/testing/_internal/distributed/rpc/jit/rpc_test.py

```diff
@@ -41,10 +52,352 @@ def forward(self):
+    str_arg: str,
+    int_arg: int,
+    tensor_kwarg: Tensor = torch.tensor([2, 2]),
+    str_kwarg: str = "str_kwarg",
+    int_kwarg: int = 2,
+):
+    return tensor_arg + tensor_kwarg, str_arg + str_kwarg, int_arg + int_kwarg
+
+
+@torch.jit.script
+def raise_script():
+    raise RuntimeError("Expected error")
+    return 0
+
+
+@torch.jit.script
+def rpc_async_call_remote_torchscript_in_torchscript(
+    dst_worker_name: str, args: Tuple[Tensor, Tensor], kwargs: Dict[str, Tensor]
+):
+    fut = rpc.rpc_async(dst_worker_name, two_args_two_kwargs, args, kwargs)
+    ret = fut.wait()
+    return ret
+
+
+class JitRpcAsyncOpTest:
+    # Call functions remotely from Script.
+    @dist_init
+    def test_all_kwargs_are_populated_by_defaults(self):
+        if self.rank != 0:
+            return
+
+        dst_worker_name = "worker{}".format((self.rank + 1) % self.world_size)
+
+        args = (torch.tensor([1, 1]), torch.tensor([2, 2]))
+        kwargs = {}
+        ret = rpc_async_call_remote_torchscript_in_torchscript(
+            dst_worker_name, args, kwargs
+        )
+        self.assertEqual(ret, torch.tensor([10, 10]))
+
+    @dist_init
+    def test_some_kwargs_are_populated_by_defaults(self):
+        if self.rank != 0:
+            return
+
+        dst_worker_name = "worker{}".format((self.rank + 1) % self.world_size)
+
+        args = (torch.tensor([1, 1]), torch.tensor([2, 2]))
+        kwargs = {"first_kwarg": torch.tensor([2, 2])}
+        ret = rpc_async_call_remote_torchscript_in_torchscript(
+            dst_worker_name, args, kwargs
+        )
+        self.assertEqual(ret, torch.tensor([9, 9]))
+
+    @dist_init
+    def test_no_kwargs_are_populated_by_defaults(self):
+        if self.rank != 0:
+            return
+
+        dst_worker_name = "worker{}".format((self.rank + 1) % self.world_size)
+
+        args = (torch.tensor([1, 1]), torch.tensor([2, 2]))
+        kwargs = {
+            "first_kwarg": torch.tensor([2, 2]),
+            "second_kwarg": torch.tensor([3, 3]),
+        }
+        ret = rpc_async_call_remote_torchscript_in_torchscript(
+            dst_worker_name, args, kwargs
+        )
+        self.assertEqual(ret, torch.tensor([8, 8]))
+
+    @dist_init
+    def test_kwargs_in_the_front_can_be_specified_by_extra_args(self):
+        if self.rank != 0:
+            return
+
+        dst_worker_name = "worker{}".format((self.rank + 1) % self.world_size)
+
+        @torch.jit.script
+        def rpc_async_call_remote_torchscript_in_torchscript_with_extra_arg(
+            dst_worker_name: str,  # noqa: E999
+        ):
+            args = (
+                torch.tensor([1, 1]),
+                torch.tensor([2, 2]),
+                # This extra arg will be fed to the first kwarg.
+                torch.tensor([2, 2]),
+            )
+            kwargs = {"second_kwarg": torch.tensor([3, 3])}
+            fut = rpc.rpc_async(dst_worker_name, two_args_two_kwargs, args, kwargs)
+            ret = fut.wait()
```

Comment:
Can we double check the `wait` here triggers the JIT continuation logic (https://github.com/pytorch/pytorch/blob/cb8d9f99aaf66947e371ae750eccd23b508fee91/torch/csrc/jit/runtime/interpreter.cpp#L1116) and doesn't actually block the JIT thread?

Reply:
This interpreter instruction, `OpCode::WAIT`, is inserted and because a symbol `aten::wait` is insert into IR gragh by the Python API, `torch.jit._wait(fut)`. 
https://github.com/pytorch/pytorch/blob/cab8772c6c8c3ce4bdd44b1889d7b28e6cacf0b0/torch/csrc/jit/python/init.cpp#L735

But I think we (training toolkit) will use `fut.wait()` more often, let me check this code path.

Reply:
you can just print the ir in your test to see if there's a `aten::wait` node for the quick check

Reply:
@wanchaol @pritamdamania 

```
graph(%dst_worker_name.1 : str,
      %args.1 : (Tensor, Tensor),
      %kwargs.1 : Dict(str, Tensor)):
  %9 : str = prim::Constant[value="__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.two_args_two_kwargs"]() # /data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/jit/rpc_fork#binary,link-tree/torch/testing/_internal/distributed/rpc/jit/rpc_test.py:185:10
  %fut.1 : Future[Tensor] = prim::rpc_async(%dst_worker_name.1, %9, %args.1, %kwargs.1) # /data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/jit/rpc_fork#binary,link-tree/torch/testing/_internal/distributed/rpc/jit/rpc_test.py:185:10
  %ret.1 : Tensor = aten::wait(%fut.1) # /data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/jit/rpc_fork#binary,link-tree/torch/testing/_internal/distributed/rpc/jit/rpc_test.py:186:10
  return (%ret.1)
```

`fut.wait()` translates to IR `ret = aten::wait(fut)`, so the first question is confirmed and positive.